### PR TITLE
MMP-224 Domain Bundle validations

### PR DESCRIPTION
**Pull request by WilliamPinho**

### CODE REVIEW

File: mule-packager/src/main/java/org/mule/tools/api/classloader/model/util/ArtifactUtils.java [hunk 1]
```diff
@@ -162,4 +163,11 @@ public static BundleDescriptor toBundleDescriptor(ArtifactCoordinates artifactCo
         .setClassifier(artifactCoordinates.getClassifier())
         .setType(artifactCoordinates.getType()).build();
   }
+
+  public static BundleDependency toBundleDependency(Artifact artifact) {
```
Comment (damiansima) [inline mule-packager/src/main/java/org/mule/tools/api/classloader/model/util/ArtifactUtils.java]: could you please add a test for this?

File: mule-packager/src/main/java/org/mule/tools/api/validation/DomainBundleProjectValidator.java [hunk 2]
```diff
@@ -33,8 +44,111 @@ public DomainBundleProjectValidator(Path projectBaseDir, Project dependencyProje
...
+
+  /**
+   * Validates that all applications refers to a unique domain.
+   *
+   * @param domain maven coordinates of the unique domain
+   * @param applications list of domain bundle applications
+   */
+  protected void validateApplications(ArtifactCoordinates domain, List<ArtifactCoordinates> applications)
+      throws ValidationException {
+    if (applications == null || applications.size() == 0) {
```
Comment (damiansima) [inline mule-packager/src/main/java/org/mule/tools/api/validation/DomainBundleProjectValidator.java]: application.size() -> application.isEmpty()

File: mule-packager/src/main/java/org/mule/tools/api/validation/DomainBundleProjectValidator.java [hunk 3]
```diff
@@ -33,8 +44,111 @@ public DomainBundleProjectValidator(Path projectBaseDir, Project dependencyProje
...
+   * @param applicationCoordinates maven coordinates of the application that is being validated
+   */
+  protected void validateApplication(ArtifactCoordinates domain, ArtifactCoordinates applicationCoordinates)
+      throws ValidationException {
+    Set<ArtifactCoordinates> applicationDomains = getApplicationDomains(applicationCoordinates);
+
+    if (applicationDomains.size() != DOMAIN_BUNDLE_VALID_NUMBER_OF_DOMAINS || !applicationDomains.contains(domain)) {
+      throw new ValidationException("Every application in the domain bundle must refer to the specified domain: " + domain
+          + ". However, the application refers to the following domain(s): "
+          + applicationDomains.stream().collect(Collectors.toList()));
```
Comment (damiansima) [inline mule-packager/src/main/java/org/mule/tools/api/validation/DomainBundleProjectValidator.java]: Can we custom the message to reflect more specifically that the application has reference to no domain?

File: mule-packager/src/main/java/org/mule/tools/api/validation/DomainBundleProjectValidator.java [hunk 4]
```diff
@@ -33,8 +44,111 @@ public DomainBundleProjectValidator(Path projectBaseDir, Project dependencyProje
...
+  protected void validateDomain(Set<ArtifactCoordinates> domains) throws ValidationException {
+    checkArgument(domains != null, "Set of domains should not be null");
+    if (!domains.stream()
+        .allMatch(artifactCoordinates -> StringUtils.equals(artifactCoordinates.getClassifier(), MULE_DOMAIN.toString()))) {
+      String message = "Not all dependencies are mule domains";
+      throw new ValidationException(message);
+    }
+    if (domains.size() != DOMAIN_BUNDLE_VALID_NUMBER_OF_DOMAINS) {
+      String message = "A mule domain bundle must contain exactly one mule domain. ";
+      if (domains.size() == 0) {
```
Comment (damiansima) [inline mule-packager/src/main/java/org/mule/tools/api/validation/DomainBundleProjectValidator.java]: domains.size() -> domains.isEmpty()

File: mule-packager/src/main/java/org/mule/tools/api/classloader/model/util/ArtifactUtils.java (hunk 1 above)
Comment (WilliamPinho) [reply mule-packager/src/main/java/org/mule/tools/api/classloader/model/util/ArtifactUtils.java]: Actually, I created that but didn't use, so I will just delete that.

File: mule-packager/src/main/java/org/mule/tools/api/validation/DomainBundleProjectValidator.java (hunk 2 above)
Comment (WilliamPinho) [reply mule-packager/src/main/java/org/mule/tools/api/validation/DomainBundleProjectValidator.java]: Done

File: mule-packager/src/main/java/org/mule/tools/api/validation/DomainBundleProjectValidator.java (hunk 4 above)
Comment (WilliamPinho) [reply mule-packager/src/main/java/org/mule/tools/api/validation/DomainBundleProjectValidator.java]: Done

File: mule-packager/src/main/java/org/mule/tools/api/validation/DomainBundleProjectValidator.java (hunk 3 above)
Comment (WilliamPinho) [reply mule-packager/src/main/java/org/mule/tools/api/validation/DomainBundleProjectValidator.java]: Done

**Pull request merged**